### ISSUE TITLE
About implement 

### ISSUE THREAD

**Issue by DavidYanAnDe** [choH/lottery_regulated_grouped_kernel_pruning#2]
Could you please provide the complete running requirement? I would like to know the specific environment configuration, such as the python version and the GPU under which the experiment is conducted, and is it necessary to use the ancient sklearn=0.20.1? Installation throws a lot of errors, both on Linux and Windows.

**Comment by DavidYanAnDe** [comment]
OK, I successfully installed the other components, but now there are errors as shown below. I would like to know the specific versions of your various packages, It feels like you're using a lot of older packages. Can you help me solve the problem in the picture? Thank you.
![U1`C8WA X1LY(E@3K76A DB](https://user-images.githubusercontent.com/48938893/203274903-07297137-eee7-4b4f-93c8-c1afba772af4.png)

**Comment by henryzhongsc** [comment]
Hi @DavidYanAnDe , 

Thanks for trying to replicate. So with distance metrics set as `cosine`, `p` is a redundant variable as it only makes sense under a euclidean/minkowski context. I believe this argument is copied from the euclidean/minkowski experiments we conducted. 

It looks like the version of `scipy` Colab used to default to (when we published this paper) has a mechanism to ignore the `p` kwargs input when not in use — as you can see our [demo Colab file](https://github.com/choH/lottery_regulated_grouped_kernel_pruning/blob/main/TMI_GKP_demo.ipynb) runs just fine without specifying a `scipy` version — but maybe Colab changes its default `scipy` version now. 

I will check the environment setting on our server, try to replicate those runs on Colab, and decide if I want to update the code or specify a `scipy` version. With a quick scan of [`pdist`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html) docs, I suspect a simple removal of `p` will do here; but let me get back to you for sure.

(And yes, we have to rely on an ancient version of `sklearn` because we depend on [ndanielsen/Same-Size-K-Means](https://github.com/ndanielsen/Same-Size-K-Means) for equal-size $k$-means.)

**Comment by DavidYanAnDe** [comment]
Thank you for your advice, I deleted the p parameter and it can be run, but there are still some problems, first of all, why the number of clusters is always 8, I don't seem to mention it in the paper, second, based on the observation of the running results and the code, your code seems to directly load a trained baseline model and test the accuracy, then perform clustering and pruning on that model, after which move on to fine-tuning, do I understand that right? In addition, could you please provide setting.json, baselinemodels and snapshot.pt files in addition to resnet20_cifar10？I am not aware of ticket windows for other experiments.

**Comment by DavidYanAnDe** [comment]
Even if I set task=train in the command json, it wouldn't train because main.py line 211 would execute load() anyway. If I want to train baseline model myself, please tell me the details of how to use it. Thank you very much.

**Comment by henryzhongsc** [comment]
Hi @DavidYanAnDe ,

First, let us make sure you can replicate our ResNet-20 demo experiment.

So I have [updated the code](https://github.com/choH/lottery_regulated_grouped_kernel_pruning/commit/fe2000965d8086600ec6235d3fe41d1b2e1c7c09) to resolve the version compatibility issue you reported on `pdist`, it is just removing the `p` parameter. It now supports the default `scipy` version on Colab (currently, it is `1.7.3`). FYI, I also checked our server, and the version we used to run on is `1.5.4`. 

I have rerun the demoed ResNet-20 experiment with the same script supplied in [demo Colab file](https://github.com/choH/lottery_regulated_grouped_kernel_pruning/blob/main/TMI_GKP_demo.ipynb) (the cell with `--exp_desc resnet20_cifar10_demo`), and it works fine. Please let me know if there's more I can help.

---

Regarding a couple of other issues you mentioned/asked, I will try to address them one by one:

> Why is the cluster number always 8? It is not mentioned in the paper.

This is a parameter we handcrafted. The intuition behind it is simple: because for the first couple layers of ResNets, the dimensions are `16, 16, 3, 3`. So 8 is the largest `n_clusters` we can use while maintaining the grouped kernel pruning format. We want to use the same group size to simplify our experiment workload, and `n_cluster: 8` performed well with `pruning_rate: 0.4375` (together, they'd yield a pruned model that's comparable to results reported in other arts).

I quickly scanned my paper, and you are probably correct that this `n_clusters` param is not explicitly reported. We never meant to hide it, as it is visible in the `setting.json` files we supplied, which we mentioned in A.3.2. of the paper. We also specified `groups = 8` in A.3.3. as an example on the issue of torch's grouped convolution implementation. Though I understand that's for a different topic, we set it as such to be consistent to our experiments. I will take note to specify it should we'd publish another GKP-related work.

> Your code seems to directly load a trained baseline model and test the accuracy, then perform clustering and pruning on that model, after which move on to fine-tuning, do I understand that right?

Yes.

> Can you provide supplement files in addition to resnet20_cifar.

Probably not, because those snapshot.pt files are huge. Even with just the 40-ish snapshots used in the demo ResNet-20 experiment, the repo is over 100MB — which IMO is already too large. Also, if you'd train your own baselines, you have no use for such files as those snapshots are only useful for their own training schedule. 

Our codebase support using the clustering info extracted from another already clustered model (we built this feature to efficiently try out different pruning strategies/settings). If you'd like, I can supply you with such clustering info, but that will only make sense if we'd cluster upon the same trained baseline. 

> I am not aware of ticket windows for other experiments.

For every 300 epochs experiment, it is `35` to `70`. You'd find most of them in `setting.json` files under [demo_experiments](https://github.com/choH/lottery_regulated_grouped_kernel_pruning/tree/main/demo_experiments) folder. I would add it honestly doesn't matter too much as there's usually a dominating clustering scheme.

> Even if I set task=train in the base json, it wouldn't train because main.py line 211 would execute load() anyway.

It is by design. You should supply it with an initialized (but yet to be trained) model at `--model_dir`.

---

Again, please reply here if you got more questions. I am happy to help :)

**Comment by DavidYanAnDe** [comment]
So I just need to provide an untrained baseline model myself and get snapshots automatically?

**Comment by henryzhongsc** [comment]
> So I just need to provide an untrained baseline model myself and get snapshots automatically?

Yes. It has been a while — and I have made another (cleaner) codebase to do more GKP works — but this `main.py` should save all epoch snapshots according to its `setting['management']['save_subdir']`. Note that when training a baseline, the initial `lr` is 0.1 instead of 0.01 for fine-tuning.

**Comment by DavidYanAnDe** [comment]
Thank you very much for your reply. I will continue to pay attention to your new work about GKP. Best wishes.

**Comment by DavidYanAnDe** [comment]
Sorry to bother you again. Could you please provide the baseline model that you seriously use? Because I see that you only have resnet code for cifar in your model folder.

**Comment by henryzhongsc** [comment]
@DavidYanAnDe What do you mean by "seriously use?" 

Also, after you asked about the snapshot saving mechanism for baseline training this morning, I went ahead and mocked some reruns, and there are indeed some changes needed in the `setting.json` file to register such snapshots. 

If you don't mind waiting a bit (should be around 2 hours-ish), I can update the demo colab file with a complete walkthrough from training the baseline to pruning and fine-tuning the model, with all supplement files supplied.

**Comment by DavidYanAnDe** [comment]
This refers to all the baseline models you used in the paper. At the moment, there is only one file in the baseline folder, resnet20_cifar10_baseline, This file is not very large.

If you provide a new json file with a complete training for baseline models, snapshot generation, and pruning, then I can try to generate baseline myself. Thank you very much for your update and reply

**Comment by henryzhongsc** [comment]
The only trained baseline provided for now is a ResNet-20 on CIFAR. This is because to run the TMI-GKP method, you will need both a target model and its snapshots as input. I can provide you other baseline models, but they won't be good for any actual use — at least not under the context of TMI-GKP — unless I also provide you their snapshot files; which, as we discussed, are probably too large to include here, yet I kinda want to keep the repo slim.

What I plan to do is I'd update the repo with a complete walkthrough of my method on ResNet-20 (model initialization, baseline training, snapshot registration, pruning, fine-tuning, with `setting.json` files for all tasks mentioned here). If you'd fancy to try it out on another network-dataset pair, you may simply change a line in the model initialization cell and rerun the whole process. 

Let me know if this is good, thanks.

**Comment by henryzhongsc** [comment]
@DavidYanAnDe I have now updated the repo (marked by commit bd8189b). The updated [Colab demo file](https://github.com/choH/lottery_regulated_grouped_kernel_pruning/blob/main/TMI_GKP_demo.ipynb) includes reusable cells on every stage of a typical TMI-GKP task. 

My previous repo design was meant to give people a quick touch on the pruning part of TMI-GKP — since that's the part we authors spent the most time on — but it didn't do a good job of explaining how to replicate a complete task cycle. The new demo (with its supplementary files) is much more self-inclusive and slim. We thank you for opening this issue, as it inspired us to improve the reproducibility and usability of our shared codebase.

Please close the issue if you consider it resolved. We wish you a happy holiday :)

**Comment by DavidYanAnDe** [comment]
Thank you very much for your update, I see all the networks about cifar from generating baselines, training, generating snapshots, pruning and fine-tuning, so for tiny-imageNet and imageNet I just need to change the --dataset in main.py during training? The other process of generating baselines, training, generating snapshots, pruning and fine-tuning is the same as cifar right?

**Comment by henryzhongsc** [comment]
@DavidYanAnDe Not exactly. You will need to change the model initialization (as they have different input sizes and `num_classes`),  `--dataset` arg, and supply tasks with `setting.json` files of correct specifications. The latter two changes apply to both baseline training and fine-tuning stage.

For example, to run Tiny-ImageNet, you should set `num_classes = 200` during model initialization. Note that Tiny-ImageNet on BasicBlock ResNets (e.g., ResNet-56 in paper) is an extra bit tricky as we used the "same" model on both CIFAR-10 and Tiny-ImageNet. Again, it has been a while so I don't remember 100% on such details, but I believe you should change line 70 in [`models/resnet.py`](https://github.com/choH/lottery_regulated_grouped_kernel_pruning/blob/main/models/resnet.py) to `self.avgpool = nn.AvgPool2d(16)` for Tiny-ImageNet; either way, it'd be a clear dim broadcasting issue addressable with a change on `avgpool` param.

In terms of having `setting.json` files with correct specifications, please refer to A.3.1 and A.3.2 of the paper. In practice and as an overview, Tiny-ImageNet experiments should share the same settings as the CIFAR experiments except for pruning rate. ImageNet experiment changes epoch budget and weight decay on top of that.

Also, it might be worth noting that you will need A100 — to my knowledge, this GPU is only available to Pro+ users — to run our Tiny-ImageNet experiments; yet you can't run ImageNet experiment on Colab at all (even with Pro+). So you'd need to offload the baseline training and fine-tune part to another environment should you want to replicate our ImageNet result.

**Comment by DavidYanAnDe** [comment]
OK, I see. Thank you for your answer. Have a nice Day :).

**Comment by henryzhongsc** [comment]
@DavidYanAnDe Cheers and likewise. As we are here with all those exchanges, do you mind I ask what you are trying to achieve here? Maybe a benchmark paper?

**Comment by DavidYanAnDe** [comment]
There is no complete idea or paper. I am still experimenting and trying. I'm still a fresh man new to pruning. Xb

**Comment by henryzhongsc** [comment]
I see. If you do end up wanting to explore the GKP realm for publications, please consider reach out to me via email. I just started my PhD, and I'd certainly appreciate more GKP adaptations. 

Godspeed :)
